### DISCUSSION TITLE
Task "Authorize all hosts' root SSH public keys" fails with error "'ansible.vars.hostvars.HostVarsVars object' has no attribute 'proxmox_root_id_rsa_pub"

### DISCUSSION THREAD

**Discussion by marcin-github** [lae/ansible-role-proxmox#211] [Q&A]
Hello!
I upgraded role from and hitted something new for me. I'm getting error:
```
TASK [lae.proxmox : Fetch root SSH public key] *****************************************************************************************************************************
ok: [xxxx.com] => {"changed": false, "content": "c3xxxxyyyyzzzzz9vdWNwcnhtNQo=", "encoding": "base64", "source": "/root/.ssh/id_rsa.pub"}
Friday 09 December 2022  10:36:24 +0100 (0:00:00.424)       0:00:03.988 *******
Friday 09 December 2022  10:36:24 +0100 (0:00:00.424)       0:00:03.987 *******

TASK [lae.proxmox : Authorize all hosts' root SSH public keys] *************************************************************************************************************
fatal: [xxxx.com]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'proxmox_root_id_rsa_pub'\n\nThe error appears to be in '/home/zzzz/ansible/roles/lae.proxmox/tasks/ssh_cluster_config.yml': line 21, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Authorize all hosts' root SSH public keys\n  ^ here\n"}

```
I tried run role for one host and for all group but this error always happens. Am I doing something wrong? 
ansible-playbook 2.10.8

**Comment by lae** [comment]
> I upgraded role from

Did you forget version numbers or something here?

> I tried run role for one host and for all group

Can you clarify this a little further? Are you just running the playbook on one host?

I believe the error you're running into would happen when you are using `--limit` or something similar to run the role on a subset of nodes, however that is not supported. Ansible needs to run the playbook on all hosts so that it can collect host keys and other information from other hosts in your cluster.

**Reply by marcin-github** [reply]
> > I upgraded role from
> 
> Did you forget version numbers or something here?

yeah;) I changed my mind how to write sentence in the middle of writing... I was using role with one year old version.

> > I tried run role for one host and for all group
> 
> Can you clarify this a little further? Are you just running the playbook on one host?
> 
> I believe the error you're running into would happen when you are using `--limit` or something similar to run the role on a subset of nodes, however that is not supported. Ansible needs to run the playbook on all hosts so that it can collect host keys and other information from other hosts in your cluster.

Yes, I'm running using --limit, I run it earlier passing all proxmox group to --limit and it worked ;) Now I try to run without --limit, for all hosts, also hosts not belongs to proxmox group and it also fails.

**Reply by lae** [reply]
Using --limit to the whole specific proxmox group (defined in role variable as `pve_group`) should also be fine. I feel like I'm still unclear as to what you're doing. Can you post the following?

- the exact command you're executing with `ansible-playbook`, with host fqdns and group names obscured if needed
- your role variables for the groups and hosts

The command output you posted earlier only showed one host in the output. That suggests that you're running the playbook against one host inside of a cluster, and this can't work at all.

**Reply by marcin-github** [reply]
command is:
`$ ansible-playbook -i inventories/ -D main.yml  -v -l xxx_pve -C -t proxmox`

variables for group:
```

pve_remove_subscription_warning: true
pve_run_system_upgrades: false
pve_run_proxmox_upgrades: false
pve_remove_old_kernels: false
pve_group: 'xxx_pve'
pve_cluster_enabled: true

pve_ceph_enabled: true
pve_ceph_network: '10.222.9.0/24'
```

when I run against group I'm also getting the same error on every host.
